### ISSUE TITLE
Polygon (ol.layer.Vector) not filled completely when used without stroke color

### ISSUE THREAD

**Issue by jbelien** [openlayers/openlayers#6341]
Hi,

I have a problem with GeoJSON file used with `ol.layer.Vector` :

The following code :
```
    var myLayer = new ol.layer.Vector({
        source: new ol.source.Vector(),
        style: new ol.style.Style({
            fill: new ol.style.Fill({
                color: [255, 0, 0, 0.2]
            }),
            zIndex: Infinity
        })
    });
```
generates this image (you can see the polygon on the left is not filled correctly) :
![capture-1](https://cloud.githubusercontent.com/assets/1150563/21803575/e95a9672-d72a-11e6-98db-fe74a0a0481e.PNG)

and everything is ok when I add a stroke color :
```
    var myLayer = new ol.layer.Vector({
        source: new ol.source.Vector(),
        style: new ol.style.Style({
            fill: new ol.style.Fill({
                color: [255, 0, 0, 0.2]
            }),
            stroke: new ol.style.Stroke({
                color: [255, 0, 0, 0.2]
            }),
            zIndex: Infinity
        })
    });
```
![capture-2](https://cloud.githubusercontent.com/assets/1150563/21803612/1b700156-d72b-11e6-90a6-c3b7614cccd9.PNG)

Any idea where the problem could come from ?
Thanks !

**Comment by ahocevar** [comment]
This is probably because of invalid polygons in the GeoJSON files. The GeoJSON spec requires the first and last coordinate of a polygon to be the same, and OpenLayers expects that as well.

**Comment by jbelien** [comment]
That was my first idea too but the object seems correct (*EPSG::3857*) :
```
{
  "type": "Feature",
  "id": "11002A05-",
  "properties": {
    
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          489499,
          6658893
        ],
        [
          489828,
          6659895
        ],
        [
          489874,
          6659829
        ],
        [
          490023,
          6659585
        ],
        [
          490056,
          6659433
        ],
        [
          489918,
          6658824
        ],
        [
          489499,
          6658893
        ]
      ]
    ]
  }
}
```

And I only display this one, it's filled completely but when OL3 has to display both, it's not filled completely.
The other one (on the right) (*EPSG::3857*) is :
```
{
  "type": "Feature",
  "id": "11002A04-",
  "properties": {
    
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          489828,
          6659895
        ],
        [
          489971,
          6659886
        ],
        [
          490249,
          6659849
        ],
        [
          490336,
          6659615
        ],
        [
          490401,
          6658929
        ],
        [
          490173,
          6658879
        ],
        [
          489945,
          6658746
        ],
        [
          489918,
          6658824
        ],
        [
          490056,
          6659433
        ],
        [
          490023,
          6659585
        ],
        [
          489874,
          6659829
        ],
        [
          489828,
          6659895
        ]
      ]
    ]
  }
}
```

Thanks 😃

**Comment by ahocevar** [comment]
I created a JSFiddle with your features, and I cannot reproduce the issue: https://jsfiddle.net/2jyxdx5a/. Can you update the fiddle so it shows the issue?

**Comment by jbelien** [comment]
Strange indeed ! I'll post a JSFiddle replicating the issue tomorrow morning !

**Comment by jbelien** [comment]
Actually, it was not easy to replicate...
But you can see the problem here : https://jsfiddle.net/jmkyp0xb/4/ (I tried to simplify the example at best but the problem does only show like this...)

The geometry of the polygon not filled completely is : 
```
{
  "type": "Feature",
  "id": "72021B032",
  "properties": {
    
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          633705,
          6639239
        ],
        [
          633961,
          6639179
        ],
        [
          633882,
          6639049
        ],
        [
          633919,
          6638935
        ],
        [
          634000,
          6638946
        ],
        [
          634072,
          6638690
        ],
        [
          634113,
          6638500
        ],
        [
          634036,
          6638437
        ],
        [
          634440,
          6637634
        ],
        [
          633852,
          6637360
        ],
        [
          633491,
          6638430
        ],
        [
          633669,
          6639244
        ],
        [
          633705,
          6639239
        ]
      ]
    ]
  }
}
```

Thanks a lot.

**Comment by ahocevar** [comment]
Thanks for the fiddle. The issue is only present with cloned features, so I suspect we do something wrong when cloning polygon geometries. I created two minimal fiddles: 
* https://jsfiddle.net/jmkyp0xb/5/: with cloning, issue present
* https://jsfiddle.net/jmkyp0xb/6/: without cloning, issue absent

**Comment by jbelien** [comment]
Awesome ! Indeed I didn't think about cloning.
Really glad you found the source of the issue. For the moment, I set a stroke color and with a stroke color the polygon is filled completely (even when cloned).

**Comment by fredj** [comment]
@jbelien it's not related to this issue but the crs in the [geojson](https://vps.geocode.be/test/bpost-gs/examples/json/nis9.json) is invalid:
```
"crs": {
  "type": "name",
  "id": "urn:ogc:def:crs:EPSG::3857"
  "properties": {}
}
```
But it should be:
```
"crs": {
  "type": "name",
  "properties": {
    "name": "urn:ogc:def:crs:EPSG::3857"
  }
}
```

see http://geojson.org/geojson-spec.html#named-crs

**Comment by jbelien** [comment]
@fredj : Thanks, good to know. the CRS was generated by `ogr2ogr`.
This specification is obsolote anyway ^^ (crs must be `urn:ogc:def:crs:OGC::CRS84` according to new specification : https://tools.ietf.org/html/rfc7946#section-4)

**Comment by fredj** [comment]
You're right ! I forgot that

**Comment by ahocevar** [comment]
This is tougher than I thought. The geometry is cloned fine. There must be something wrong in the renderer. Investigating this now.

**Comment by jbelien** [comment]
Thanks @ahocevar ! Cheer up !

**Comment by ahocevar** [comment]
Pull request created: #6358.

**Comment by jbelien** [comment]
Hi @ahocevar , thnaks for the pull request.
I cloned the repository and build the `ol.js` (`// Version: v3.20.0-168-g3c751ca`) but I think there still is an issue :
![capture-3](https://cloud.githubusercontent.com/assets/1150563/21884709/b6428962-d8b4-11e6-9e98-a55d8660fff1.PNG)

**Comment by ahocevar** [comment]
Looks like you have built a wrong revision. The correct version is `v3.20.0-168-ga671687`. Can you please try [this `ol.js`](https://dl.dropboxusercontent.com/u/5489321/ol.js)?

**Comment by jbelien** [comment]
So sorry, I didn't check that #6358 was not yet merged.
I confirm this issue is fixed with #6358 👍 
Awesome work, thank you so much ! A fresh beer is waiting for you in Brussels \o/
